akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala

```diff
@@ -65,13 +65,23 @@ private[http] object WebSocketClientBlueprint {
         new GraphStageLogic(shape) with InHandler with OutHandler {
           // a special version of the parser which only parses one message and then reports the remaining data
           // if some is available
-          val parser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings, log)) {
+          val parser: HttpResponseParser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings, log)) {
             var first = true
             override def handleInformationalResponses = false
             override protected def parseMessage(input: ByteString, offset: Int): StateResult = {
               if (first) {
-                first = false
-                super.parseMessage(input, offset)
+                try {
+                  // If we're called recursively then that's a next message
+                  first = false
```

Comment:
Would it also work to just switch the two statements?

```scala
val result = super.parseMessage(input, offset)
first = false
result
```

Reply:
No, apparently `parseMessage` recursively calls itself (and then you want 'first' to be false for the recursive call) - I haven't looked into how/why in detail but that's what happened :smile:

This is a risk: what if `NotEnoughDataException` is thrown while processing the next element in that recursive call? Since the recursive call will just does `emit(RemainingBytes(input.drop(offset))); terminate()` I don't think that can happen though.